collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadListActivity.kt

```diff
@@ -0,0 +1,708 @@
+import org.odk.collect.android.formentry.RefreshFormListDialogFragment
+import org.odk.collect.android.formentry.RefreshFormListDialogFragment.RefreshFormListDialogFragmentListener
+import org.odk.collect.android.formmanagement.FormDownloadException
+import org.odk.collect.android.formmanagement.FormDownloader
+import org.odk.collect.android.formmanagement.FormSourceExceptionMapper
+import org.odk.collect.android.formmanagement.ServerFormDetails
+import org.odk.collect.android.formmanagement.ServerFormsDetailsFetcher
+import org.odk.collect.android.fragments.dialogs.FormsDownloadResultDialog
+import org.odk.collect.android.fragments.dialogs.FormsDownloadResultDialog.FormDownloadResultDialogListener
+import org.odk.collect.android.injection.DaggerUtils
+import org.odk.collect.android.listeners.DownloadFormsTaskListener
+import org.odk.collect.android.listeners.FormListDownloaderListener
+import org.odk.collect.android.network.NetworkStateProvider
+import org.odk.collect.android.tasks.DownloadFormListTask
+import org.odk.collect.android.tasks.DownloadFormsTask
+import org.odk.collect.android.utilities.ApplicationConstants
+import org.odk.collect.android.utilities.AuthDialogUtility
+import org.odk.collect.android.utilities.AuthDialogUtility.AuthDialogUtilityResultListener
+import org.odk.collect.android.utilities.DialogUtils
+import org.odk.collect.android.utilities.WebCredentialsUtils
+import org.odk.collect.android.views.DayNightProgressDialog
+import org.odk.collect.androidshared.ui.DialogFragmentUtils.dismissDialog
+import org.odk.collect.androidshared.ui.DialogFragmentUtils.showIfNotShowing
+import org.odk.collect.androidshared.ui.ToastUtils.showShortToast
+import org.odk.collect.forms.FormSourceException
+import org.odk.collect.forms.FormSourceException.AuthRequired
+import timber.log.Timber
+import java.io.Serializable
+import java.net.URI
+import java.util.Locale
+import javax.inject.Inject
+import kotlin.collections.set
+
+/**
+ * Responsible for displaying, adding and deleting all the valid forms in the forms directory. One
+ * caveat. If the server requires authentication, a dialog will pop up asking when you request the
+ * form list. If somehow you manage to wait long enough and then try to download selected forms and
+ * your authorization has timed out, it won't again ask for authentication, it will just throw a
+ * 401
+ * and you'll have to hit 'refresh' where it will ask for credentials again. Technically a server
+ * could point at other servers requiring authentication to download the forms, but the current
+ * implementation in Collect doesn't allow for that. Mostly this is just because it's a pain in the
+ * butt to keep track of which forms we've downloaded and where we're needing to authenticate. I
+ * think we do something similar in the instanceuploader task/activity, so should change the
+ * implementation eventually.
+ *
+ * @author Carl Hartung (carlhartung@gmail.com)
+ */
+class FormDownloadListActivity :
```

Comment:
Usually I would encourage you to convert classes that you touch to kotlin bu in this case I feel uneasy. The class is big and I think it makes it risky. Please just focus on replacing asynctask with coroutines and revert this change.

Reply:
Done. I moved logic of task to viewModel and operate there with coroutines and livedata.